### PR TITLE
File name sanitation

### LINKED ISSUE #591: Exporting data items with forward slashes results in wrong filename
Forward slashes aren't allowed in Windows filenames.

Exporting a single file with forward slashes works on macOS. It is not allowed (some versions of Qt?) on Windows.

Exporting a batch with forward slashes in the prefix silently fails on both macOS and Windows.

Also, how are backslashes and other disallowed characters handled?

Is there any way to escape a filename and remove disallowed characters that works cross platform?

---

Edit 2024-05-14: This issue only applies to batch exports or possibly the default file name used when exporting a single file. In those places, we should ensure that if we generate a filename from a title, it is sanitized for use as a filename. See #874 for some initial, abandoned work. We can assume that a filename the user chooses a filename in a save file dialog is sanitized; it may still fail but will be reported to the user.

---

```[tasklist]
### Tasks
- [ ] Propose an algorithm to sanitize an arbitrary title string
- [ ] Consider platform specific restrictions, unicode, special characters including slashes, spaces, parethensis, etc.
- [ ] Consider restrictions on the length, i.e. limit to a length of 64?
- [ ] Implement algorithm, include tests (input strings -> expected output strings)
- [ ] Integrate algorithm into export dialog where we propose the initial title
- [ ] Integrate algorithm into export dialog where we allow user to export group of items
- [ ] Export batch code is in `do_export` in `ExportDialog.py`
- [ ] For extra credit, move `do_export` functionality into an `Action` object in `DocumentController.py` for scripting
```

### PR DESCRIPTION
**Pull request by KRLango**
Added a sanitation of filenames to the export.

Fixes #591

### CODE REVIEW

File: nion/swift/ExportDialog.py [hunk 1]
```diff
@@ -28,6 +29,8 @@
 from nion.utils import Geometry
 from nion.utils import Model
 
+
```
Comment (cmeyer) [inline nion/swift/ExportDialog.py:32]: Can you remove this white space change?

File: nion/swift/model/Utility.py [hunk 2]
```diff
@@ -432,6 +433,50 @@ def do_sample() -> None:
     threading.Thread(target=do_sample).start()
 
 
+def simplify_filename(filename: str, replacement_char: str = '_', maximum_length: int = 64) -> str:
```
Comment (cmeyer) [inline nion/swift/model/Utility.py:436]: I'm questioning the limit of 64 - this seems too short and maybe unnecessary. Is there a reason for this? macOS limit is 255 utf-8 characters. Windows has different restrictions, seemingly 32K when using Unicode calls (see https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry). Linux is more flexible.

File: nion/swift/model/Utility.py [hunk 3]
```diff
@@ -432,6 +433,50 @@ def do_sample() -> None:
...
+    # Windows illegal characters = < > : / \ | ? * " ASCII values 0-31 (non-printable chars)
+    # Windows illegal filenames = CON PRN AUX NUL COM1 COM2 COM3 COM4 COM5 COM6 COM7 COM8 COM9
+    #                             LPT1 LPT2 LPT3 LPT4 LPT5 LPT6 LPT7 LPT8 LPT9
+    # those files names are illegal with or without a suffix, upper and lower case
+    # Windows file names cannot start or end with periods or spaces
+
+    assert maximum_length > 5
+    assert len(replacement_char) == 1
+
+    # since windows is the most restrictive and covers the others, just focus on that
```
Comment (cmeyer) [inline nion/swift/model/Utility.py:449]: I'm not sure whether I like this idea -- but we could make the restrictions platform specific. The downside is one could save a quirky file on macOS and not be able to load it on Windows - but that may be preferable to strange limitations on macOS for macOS-only users. Any thoughts?

**Review by cmeyer — CHANGES_REQUESTED** [review]
Overall this looks good - my main concern is the 64 char limit. Also a minor whitespace issue.

File: nion/swift/model/Utility.py (hunk 2 above)
Comment (KRLango) [reply nion/swift/model/Utility.py:436]: That would be the third task in #591

Comment (KRLango) [reply nion/swift/model/Utility.py:436]: Also this is only the file name - it does not include the full path so we need to consider what a sensible maximum might be. That said, it is easy to change the default value once we make a decision.

**Pull request closed without merging**